Globalize: Add (Unicode Bidi Algorithm) compliant Bidi engine.

ashensis opened the pull request:
As per https://github.com/jquery/globalize/issues/542 discussion I submit the Bidi engine based on last Unicode Bidi character classification generated from "https://github.com/mathiasbynens/unicode-8.0.0"

ashensis commented:
I do not understand what is wrong with CLA signature, I signed up multiple times with name: ashensis and emal: sashash@il.ibm.com and still this CLA check fails. My submission (by ashensis) should match the CLA signature.

jzaefferer commented:
[The check results page](http://contribute.jquery.org/CLA/status/?repo=globalize&sha=07a040f757b29fd35cf373b9d152e3ecadd0af24) says that the name on the signature doesn't match the name on the commit. Looks to me like your signature is fine, but you're using a short/nickname on the commit. You can find instructions for updating the commit on the same page, though we can also take care of updating the commit author when merging your PR, since there is a valid signature already.

ashensis commented:
Indeed I use the short name "ashensis" for my submissions, which constitutes my ID with jQuery. I have feeling that once I got signed CLA with my full name and the same email, my further attempts to sign with short name and the same email fail. I would prefer to use "ashensis" name for my further contributions. Could you please help me with resetting the CLA to "ashensis" with the same email sahsash@il.ibm.com. Thanks in advance.

ashensis commented:
Thank you jzaefferer, I followed your advice and updated the submission author

tomerm commented:
@rxaviers did you have a chance to review this submission ? Thx.

rxaviers commented:
Hi, not yet sorry for the delay. Will do no longer than this coming week.

rxaviers commented on .gitignore:
```diff
@@ -1,5 +1,12 @@
+*.sw*
+*.gem
+.rbx
+.ruby-version
```
These files are not included by Globalize, therefore they shouldn't be listed here.

rxaviers commented on .travis.yml:
```diff
@@ -1,9 +1,25 @@
+    - env: RAILS_4_2=true DB=mysql
+notifications:
+  webhooks:
+    - https://webhooks.gitter.im/e/40dc875af26f64f39090
```
?

rxaviers commented:
@ashensis @tomerm please could you double check if this PR is what you intended to do? It's seems a completely different project has been submitted. It's ruby.

tomerm commented:
@rxaviers I am not sure about all files. I am sure Alex can explain better. But at least those 3 are definitely relevant to Bidi engine: 

src/bidi-engine.js
src/bidi/unicode-types.js
src/build/intro-bidi-engine.js

ashensis commented:
Indeed, as @Tomer said , these 3 files + Gruntfile.js corresponding changes are relevant. As for discrepancies in .gitignore .travis.yml AUTHORS.txt CONTRIBUTING.md package.json README.md and few others (fromtest\" I really don't remember how did the crall in, they most certainly have been changed with my last git pull synchronization, my apologies. Now got them updated to correct versions.

rxaviers commented on src/bidi-engine.js:
```diff
@@ -0,0 +1,473 @@
+				cType = "N";
+				break;
+		}
+		return cType;
```
Are the above rules available in CLDR somewhere? I'm wondering if they are perhaps available in XML only.

rxaviers replied:
For clarity, the above sounds to me as the "[plural](https://github.com/unicode-cldr/cldr-core/blob/master/supplemental/plurals.json)" rules for bidi.

rxaviers replied:
Nevermind, the above rules are for cTypes not locale parts.

ashensis replied:
Above mentioned code snippet represents an algorithmic part, not be expected in CLDR

rxaviers replied:
Yeap, I realized that. Nevermind.

rxaviers commented:
Codewise, could you please include unit and functional tests? For now, a couple of unit tests are sufficient to help me understand the use cases / how this feature is used. Other tests could be included later along with the review. Feel free to push new commits including it incrementally.

rxaviers commented:
Documentation is another piece I'm looking for: please update README.md including the module (and feel free to create `doc/*` files as needed). Just like the item above, feel free to push new commits incrementally implementing this.

rxaviers commented:
Is this implementation (or part of it) based on any public specification like http://unicode.org/reports/tr9/?

rxaviers commented:
There are coding styles changes to be made, but I don't want to nitpick for now. The above points are the major ones I see for now.

Thanks for your work so far.

tomerm commented:
@rxaviers Indeed, this implementation is compliant with UBA (http://unicode.org/reports/tr9/).

rxaviers commented:
> @rxaviers Indeed, this implementation is compliant with UBA (http://unicode.org/reports/tr9/). 

Excellent, so let's include source comments referencing parts of the specification where possible.

ashensis commented:
Rafael, I will try to attend all requests above in near feature, as for contributing to documentation folder we have to make sure that this doesn't contradict IBM license. We have some concerns and are planning to clarify this.

ashensis commented:
Hi Rafael, I got committed the following updates as requested:
1. Add documentation doc\api\bidi-engine\bidi-engine.md and updated Readme.md with relevant info
2. Add unites test\unit\bidi-engine\reordering.js
BTW, I don't realize the difference between unitest and functional test, so I pushed whole extensive testing under unites.
3. Update source file with appropriate references on relevant parts of http://unicode.org/reports/tr9/
Could you please have a look.

rxaviers commented on README.md:
```diff
@@ -364,6 +364,24 @@ handle dependencies and CLDR loading manually yourself.
+  // > "123&C<B>Aa(b)c" (upper case stands for Arabic or Hebrew)
+
+  [Read more...](doc/api/bidi-engine/bidi-engine.md)
+
```
Something is wrong with the markup. Click 'View' above and you'll note. I think it's a minor \`.

ashensis replied:
Can't wrap my head around, could you clarify please, what is wrong?

Arkni replied:
@ashensis here are some notes:

1. Headers

  this:
  ```md
  - ** Globalize.bidiEngine( opt ) `**
  ```
  should be:
  ```md
  - **`Globalize.bidiEngine( opt )`**
  ```
2. JS snippets:

  this
  ```
  ```javascript
  var bidiEngine = Globalize.bidiEngine( options );
  ```
  should be:
  <pre>
  ```javascript
  var bidiEngine = Globalize.bidiEngine( options );
  ```
  </pre>
  The same thing for the other snippet.

ashensis replied:
Thanks a lot for clarification, @Arkni

rxaviers commented on doc/api/bidi-engine/bidi-engine.md:
```diff
@@ -0,0 +1,84 @@
+
+bidiEngine.setOptions( options );
+
+var ret = bidiEngine.doBidiReorder( "Input text" );
```
Something is wrong with this markup too. Click 'View' above and you'll note. Perhaps, another minor thing.

Arkni replied:
The same as the above comments. Add \`\`\` in a new line after `var ret = b ...`

rxaviers commented on Gruntfile.js:
```diff
@@ -509,6 +509,25 @@ module.exports = function( grunt ) {
+									endFile: "src/build/outro.js"
+								}
+							}
 						}
```
There's a task called `uglify`, please include bidi-engine there. You can use one of the modules, e.g., date as baseline.

rxaviers commented:
Thanks for the updates. The overall is clearer to me now. Will go through the details with more time. Sorry for the delay.

Friendly reminder for @tomerm: any updates for https://github.com/jquery/globalize/issues/542#issuecomment-171463089, thanks.

rxaviers commented on src/bidi-engine.js:
```diff
@@ -0,0 +1,492 @@
+	};
+/**
+ * for reference see 3.3 in http://unicode.org/reports/tr9/
+*/
```
Some coding styles nitpick... We only use `/**/` style for headers, even long comments use `//` http://contribute.jquery.org/style-guide/js/#comments. Please, could you update these comments using `//` instead. Also indent them according to the below code.

rxaviers commented on src/bidi-engine.js:
```diff
@@ -0,0 +1,492 @@
+	var _computeLevels = function( chars, levels, params ) {
+		var len = chars.length, prevState, newState = 0, newLevel, action, condition,
+			stateTable = _dir ? _STATE_TABLE_RTL : _STATE_TABLE_LTR,
+			condPos = -1, i, idx, types = [], resolvedTypes = [];
```
Please, can you keep each variable in its own line sorted alphabetically? You could group all the ones that aren't initialized (e.g., `var action, condition, i, idx, ..., prevState, ...` ok to be in the same line)

rxaviers replied:
Is `idx` `index`? Can we use the full `index` name?

rxaviers commented on src/bidi-engine.js:
```diff
@@ -0,0 +1,492 @@
+		}
+		for ( idx = 0; idx < len; idx++ ) {
+			prevState = newState;
+			resolvedTypes[idx] = _resolveCharType( chars, types, resolvedTypes, idx );
```
Note spaces `resolvedTypes[idx]` -> `resolvedTypes[ idx ]`

rxaviers commented on src/bidi-engine.js:
```diff
@@ -0,0 +1,492 @@
+ * - isSymmetricSwapping [boolean] (defaults to false):
+ * allowed values true(needs symmetric swapping), false (no need in symmetric swapping),
+ */
+	this.$bidiEngine.setOptions = function( options ) {
```
Why `bidiEngine` has a `$` in the name please?

ashensis replied:
I wanted to signify the exposed scope by this, if you don't like this, I can remove '$'

rxaviers commented on src/bidi-engine.js:
```diff
@@ -0,0 +1,492 @@
+	return this.$bidiEngine;
+};
+
+Globalize._bidiUnicodeTypes = bidiUnicodeTypes;
```
Is `Globalize._bidiUnicodeTypes` accessed outside of this file?

ashensis replied:
Sorry, will remove from Globalize scope

rxaviers commented on test/unit/bidi-engine/reordering.js:
```diff
@@ -0,0 +1,133 @@
+  var reorderedString = bidiEngine.doBidiReorder( sourceString , sourceToTargetMap, levels );
+  assert.equal( reorderedString, "\u05d0>\u05d5<\u05ea& 321c)b(a" );
+  assert.deepEqual( sourceToTargetMap, [5,6,7,8,9,10,11,14,13,12,4,3,2,1,0] );
+  assert.deepEqual( levels, [0,0,0,0,0,1,1,1,1,1,1,1,2,2,2] );
```
Spaces `[0, 0, 0, ...`

Arkni commented on doc/api/bidi-engine/bidi-engine.md:
```diff
@@ -0,0 +1,84 @@
+	isSymmetricSwapping: true
+  };
+  var bidiEngine = Globalize.bidiEngine( options );
+
```
You should always close the snippets. Add \`\`\` in a new line after `var bindingEngine = ...`

Arkni commented on doc/api/bidi-engine/bidi-engine.md:
```diff
@@ -0,0 +1,84 @@
+  // reorderedString > "123 &C<B>Aa(b)c" (upper case stands for Arabic or Hebrew)
+  // sourceToTargetMap >
+  // levels >
+
```
The same as the above notice: you need to add \`\`\` to the next line after `// levels >`

Arkni commented on src/bidi-engine.js:
```diff
@@ -0,0 +1,492 @@
+	var _init = function( text, sourceToTargetMap ) {
+		if ( sourceToTargetMap ) {
+			for ( var i = 0; i < text.length; i++ ) {
+				sourceToTargetMap[i] = i;
```
spaces
`sourceToTargetMap[i]` => `sourceToTargetMap[ i ]`

ashensis commented:
In last commit I tried to attend all remarks as follows:
1. added bidi engine to uglify task to produce minimal version
2. replaced idx by index everywhere, changed variables declaration, 
added spaces before and after content of square brackets.
3. removed _bidiUnicodeTypes from Globalize scope, removed '$' from exposed namespace.
4. added spaces in resultant arrays in reordering.js unitest module.
5. changed /* comments with //
6. changed markup in document files

tomerm commented:
@rxaviers Any outstanding comments on this PR ?

rxaviers commented:
Hi, @jzaefferer and I have both reviewed the PR.

The diagnostic we have is similar to the one for the Arabic Shaping feature. The implementation helped us to understand the dependencies the module has on Globalize and it turns out there is almost none. It means the module is loosely coupled with Globalize. With that in mind, we suggest you create a standalone BiDi engine library. You should be able to take the code from this PR, which is already complete, and provide the bits of infrastructure around that you actually need (which is likely less then what Globalize needs), so I believe the investment you made so far by implementing this feature can be fully utilized in a standalone library. You can then update that module independently of Globalize. And it will be easier to maintain, since you won’t have to support globalize-compiler (which doesn’t provide any benefit for this engine).

Having said that, if you want to make this a jQuery hosted project as well, I believe @kborchers will be happy in assisting you further too.

rxaviers commented:
Please, just let us know if you have additional questions.

tomerm commented:
@rxaviers thanks a lot for assistance and advise. 

@kborchers I believe it does make sense to have Bidi engine library (JS implementation of Unicode Bidi Algorithm - http://unicode.org/reports/tr9/) to be a JQuery hosted projects. I would appreciate your guidance for next steps. Many thanks in advance.

kborchers commented:
@tomerm - as we work to finalize a new technical organization and support structure for the Foundation, we are defining policies, processes and procedures for current and new projects at https://github.com/jquery-foundation/TAC. I would recommend reading through that repo and then do a couple of things:

1. Let us know if anything needs clarification, more detail, etc by filing issues
2. Follow the new project application process outlined in the Project-Lifecycle.md and again, if things are unclear or need further explanation, please file issues to help us improve those documents

Also, please keep in mind that we are still finalizing that structure so some parts are not completely in place like the members of the TAC for example. We hope to bring all of that together soon but until then, there will likely be delays in working projects into the Foundation but we hope to continually improve and streamline that process as we go. This should not discourage you from starting this new project under your personal GitHub org to keep everything moving and then work toward joining the Foundation. Thanks!
